feat: cache host capability.

jvanz opened the pull request:
## Description

Adds an RFC with a proposal on how to add a new host capability to allow policy authors to store arbitrary data into a cache backend.


Fix https://github.com/kubewarden/policy-server/issues/883

jvanz committed:
feat: cache host capability.

Adds an RFC with a proposal on how to add a new host capability to allow
policy authors to store arbitrary data into a cache backend.

Signed-off-by: José Guilherme Vanz <jguilhermevanz@suse.com>

flavio commented on rfc/0024-cache-host-capability.md:
```diff
@@ -0,0 +1,199 @@
+
+## policy-evaluator Changes
+
+1. A new callback handler module will be added under `src/callback_handler`
+   containing the functions to perform the `set` and `get` cache operations.
+2. The `CallbackHandler` struct and its builder (`CallbackHandlerBuilder`) will
+   be updated with a new `cache` field. This allows `policy-server` to inject
+   the active cache backend into the handler.
+3. The `CallbackRequestType` enum will be extended to include variants for the
+   new cache operations (`CacheSet` and `CacheGet`).
+4. The `handle_request` function in `CallbackHandler` will add two new match
+   arms to route these new request types to the appropriate cache operation
+   functions.
+5. Existing host capabilities that currently use their own caching (via the
+   `cached` macro) must be refactored to use this new shared cache backend.
+   This consolidates caching logic into a single mechanism.
+6. To prevent key collisions between internal Kubewarden caches and
+   policy-defined caches, a reserved prefix (e.g., `kubewarden_internal_`) must
+   be used for all keys set by host capabilities. The `cache.set` operation must
+   reject any policy-provided key that uses this reserved prefix.
+7. The `cache` field in `CallbackHandler` will be a trait object (e.g.,
+   `Box<dyn Cache>`) to support multiple backend implementations at runtime.
+   The `cached` crate's traits can serve as a foundation for this contract.
```
In the future, don't add low level implementation details to the RFC.

flavio commented on rfc/0024-cache-host-capability.md:
```diff
@@ -0,0 +1,199 @@
+
+## policy-server Changes
+
+The `policy-server` will be responsible for instantiating the cache backend and
+passing it to the `CallbackHandlerBuilder`. The `new_from_config` function will
+be modified to create the appropriate cache backend based on runtime
+configuration. If no backend is specified, it will default to the in-memory
+cache provided by the `cached` crate. The default cache backend from cached
+crate should be `ExpiringValueCache`. It's a in memory cache that allow
+defining ttl for each value individually.
+
+For in-memory cache, it's also necessary to start a Tokio task to clean expired
+items in memory. Otherwise, we can leave expired items in memory forever if
+they are not get in future calls.
+
+Even using `cached` crate as the library manage the cache. The interaction with
+it should be abstract by some types under control of the Kubewarden team.
+Therefore, if necessary to move away of the dependency, the under the hood
+library can be changed with mimimal change in the application logic. This also
+allow the team adding another backend that is not nativaly supported by the
+`cached` crate. The goal is just to no couple Kubewarden code too much to the
+`cached` crate.
```
same here

flavio commented on rfc/0024-cache-host-capability.md:
```diff
@@ -0,0 +1,199 @@
+
+# Alternatives
+
+To avoid the performance overhead of runtime polymorphism (`dyn Trait`), we
+could define the cache backend at compile time using generics. However, this is
+not feasible for our use case, as the desired cache backend must be
+configurable at runtime based on user-provided settings.
+
+Considered not rely on `cached` crate. Even the crate seems to be used by a
+good number or person. Its repository looks missing some good practices like
+proper tagging and documentation could be better. However, writing the cache
+from scratch does not sound worth now. Furthermore, the crate is already in use
+by the project, the first iteration should continue to use it. But adding a
+abstraction layer on top of it. Just to make replacement and extension using
+other libraries easier in the future.
```
This is again a bit too low level

flavio reviewed (approved):
I like the RFC.

In some parts this gets too low level and implementation specific. The purpose of a RFC is to set the director and define the high level architecture. The implementation details are figured out when creating EPICS or breaking them down.

viccuad reviewed (approved):
LGTM, good write-up! I'm ok with the low-level bits, provided background that I lacked.

I say change the state to ACCEPTED and merge!

jvanz committed:
fix: create a section to have the details of implementation.

As mentioned during the review process. This RFC has some low level
details about the implementation of the host capability. This is not
bad, but it's optional. Therefore, I've moved the suggested
implementation details to a dedicated section.

The status of the RFC has been updated to "ACCEPTED".

Signed-off-by: José Guilherme Vanz <jguilhermevanz@suse.com>

jvanz commented:
@viccuad @flavio Thanks for the review! I've understand the comments. Thanks for them! I've moved the more in details text to a dedicated section and state it as "suggested implementation". As I do not change the content, I'm considering this PR accepted and I'll merge it now. 

If you want more changes, let me know. I'm happy to open another PR to change the RFC again.
